pkg/binpack/efficiency.go

```diff
@@ -0,0 +1,156 @@
+// is 0.81. One packing efficiency is deemed less efficient when its avg efficiency is lower than
+// the other's packing efficiency.
+func (p *AvgPackingEfficiency) LessThan(o AvgPackingEfficiency) bool {
+	return p.Max < o.Max
+}
+
+// WorstAvgPackingEfficiency returns a representation of a failed bin packing. Each individual resource
+// type is at worst possible (zero) packing efficiency.
+func WorstAvgPackingEfficiency() AvgPackingEfficiency {
+	return AvgPackingEfficiency{
+		CPU:    0.0,
+		Memory: 0.0,
+		GPU:    0.0,
+		Max:    0.0,
+	}
+}
+
+// PackingEfficiency represents result packing efficiency per resource type for one node. Computed
+// as the total resources used divided by total capacity.
+type PackingEfficiency struct {
+	NodeName string
+	CPU      float64
+	Memory   float64
+	GPU      float64
+}
+
+// Max returns the highest packing efficiency of CPU and Memory.
+func (p *PackingEfficiency) Max() float64 {
+	return math.Max(p.GPU, math.Max(p.CPU, p.Memory))
+}
+
+// ComputePackingEfficiencies calculates utilization for all provided nodes, given the new reservation.
+func ComputePackingEfficiencies(
+	nodeGroupSchedulingMetadata resources.NodeGroupSchedulingMetadata,
+	reservedResources resources.NodeGroupResources) map[string]*PackingEfficiency {
+
+	nodeEfficiencies := make(map[string]*PackingEfficiency, 0)
+
+	for nodeName, nodeSchedulingMetadata := range nodeGroupSchedulingMetadata {
+		nodeEfficiencies[nodeName] = computePackingEfficiency(nodeName, *nodeSchedulingMetadata, reservedResources)
+	}
+
+	return nodeEfficiencies
+}
+
+func computePackingEfficiency(
+	nodeName string,
+	nodeSchedulingMetadata resources.NodeSchedulingMetadata,
+	reservedResources resources.NodeGroupResources) *PackingEfficiency {
+
+	nodeReservedResources := nodeSchedulingMetadata.SchedulableResources.Copy()
+	nodeReservedResources.Sub(nodeSchedulingMetadata.AvailableResources)
+	if reserved, ok := reservedResources[nodeName]; ok {
+		nodeReservedResources.Add(reserved)
+	}
+	nodeSchedulableResources := nodeSchedulingMetadata.SchedulableResources
+
+	// GPU treated differently because not every node has GPU
+	gpuEfficiency := 0.0
+	if nodeSchedulableResources.NvidiaGPU.Value() != 0 {
+		gpuEfficiency = float64(nodeReservedResources.NvidiaGPU.Value()) / float64(normalizeResource(nodeSchedulableResources.NvidiaGPU.Value()))
+	}
+
+	return &PackingEfficiency{
+		NodeName: nodeName,
+		CPU:      float64(nodeReservedResources.CPU.Value()) / float64(normalizeResource(nodeSchedulableResources.CPU.Value())),
+		Memory:   float64(nodeReservedResources.Memory.Value()) / float64(normalizeResource(nodeSchedulableResources.Memory.Value())),
+		GPU:      gpuEfficiency,
+	}
+}
+
+func normalizeResource(resourceValue int64) int64 {
+	if resourceValue == 0 {
+		return 1
+	}
+	return resourceValue
+}
+
+// ComputeAvgPackingEfficiency calculate average packing efficiency for an entire cluster, given
+// packing efficiencies for individual nodes.
```

Comment:
This is just for a group of nodes, not the cluster right?

Reply:
Absolutely. Fixed.